### PR TITLE
Enable more CI tests for k8s e2e

### LINKED ISSUE #1190: Network Policy feature Regression in the master branch
@dcbw @danwinship @dceara  looks like we have regressed NetworkPolicy feature in the master.

This is the commit that has regressed the repo.
--------8<---------8<-------
commit 591245645e8313dd458525bb8d2b2c952ff230ce
Author: Dumitru Ceara <dceara@redhat.com>
Date:   Wed Mar 25 14:15:22 2020 +0100

    ovn: Fix default deny ACLs.
    
    Deny ACLs for PolicyTypeEgress policies should be applied in direction
    "from-lport". Also, make the "find ACL" ovn-nbctl calls more specific by
    matching on direction and priority too.
    
    Fixes #1174.
    
    Signed-off-by: Dumitru Ceara <dceara@redhat.com>
--------8<---------8<-------

We cannot use `from-lport` direction as the ACLs are applied on the `ingress` pipeline of the switch even before LB has done its job. So, this will fail for Cluster IPs. See PR #370  for more info.
We will need to revert his commit. Thanks.

### PR DESCRIPTION
**Pull request by trozet**
A bunch of network policy tests were being skipped which should pass now.

Closes #1190

Signed-off-by: Tim Rozet <trozet@redhat.com>

### CODE REVIEW

**Comment by trozet** [comment]
@dcbw @danwinship PTAL at this. It passes the named port k8s tests I ran it against, but I still have some questions about Named Port locality. Meaning that, I'm wondering if the names are unique across namespace? I'm guessing they are not unique globally, although the current code treats it that way. I was just trying to get something working for now. I don't see much on the k8s docs about named ports so I'm looking for some help here. Thanks.

You can see I'm basically leveraging the port cache to store the named ports found during pod add, then looking up the value when we see a network policy created with a named port.

**Comment by danwinship** [comment]
so the idea of named ports is that you can make a policy that says "allow from namespace 'prometheus' to the port named 'metrics' on any pod", and then if you have Pod A with port 9999 named 'metrics' and Pod B with port 8888 named 'metrics', then the policy allows traffic to port 9999 on Pod A and port 8888 on Pod B. So every pod that matches may have the same or a different port than every other pod that matches.

**Comment by trozet** [comment]
thanks @danwinship so it sounds like the way I store them just global names to values in the port cache is fine, but when I go to search for a port name, i need to return all of them instead of just the first one I find, and then add all applicable ports to the policy.

File: go-controller/pkg/ovn/port_cache.go [hunk 1]
```diff
@@ -90,3 +93,14 @@ func (c *portCache) remove(logicalPort string) {
 		}
 	}()
 }
+
+func (c *portCache) getValueOfNamedPort(namedPort string) (int32, error) {
```
Comment (trozet) [inline go-controller/pkg/ovn/port_cache.go:97]: I think this needs to return a list of all the integer ports for a named port as there could be multiple

**Comment by danwinship** [comment]
OK, the NetworkPolicy code would have to be restructured a bunch to make this work. Currently it assumes it can think about the ports independently of everything else, but that stops being true with named ports. In particular, the assumption that each `gressPolicy` maps to a single ACL is no longer correct.

For an ingress `gressPolicy` with a named port, you need to either:
1. create a separate ACL in the `namespacePolicy`'s port group for {each pod (in the namespace of the NetworkPolicy) which is matched by `policy.spec.podSelector` and has a port of the given name}, which allows traffic only to that pod IP and port, from the peers selected by the `gressPolicy` as normal, OR
2. create a separate port group for {each unique port number which is the value of the named port of some pod (in the namespace of the NetworkPolicy) which is matched by `policy.spec.podSelector`} and add pods to the correct port group for their port number, and add a single ACL to each port group which allows traffic to the corresponding port number, from the peers selected by the `gressPolicy`

So given
- Pod A with IP 10.0.0.1 and label "matching=true" and port "foo" on 1111
- Pod B with IP 10.0.0.2 and label "matching=true" and port "foo" on 2222
- Pod C with IP 10.0.0.3 and label "matching=true" and port "foo" on 2222
- Pod D with IP 10.0.0.4 and label "matching=FALSE" and port "foo" on 1111
- Pod E with IP 10.0.0.5 and label "matching=true" and no port "foo"
- a NetworkPolicy

      spec:
        podSelector:
          matchLabels:
            matching: true
        ingress:
        - from:
          - namespaceSelector:
              matchLabels:
                something: Z
          - ports:
            - port: foo

Then solution 1 would have a single port group X, containing Pod A, Pod B, Pod C, and Pod E (ie, everything matching the `spec.podSelector`, just like normal), and would contain 3 allow ACLs with `match` clauses something like:
- `ip.src == {Z} && ip.dst == 10.0.0.1 && tcp && tcp.dst == 1111`
- `ip.src == {Z} && ip.dst == 10.0.0.2 && tcp && tcp.dst == 2222`
- `ip.src == {Z} && ip.dst == 10.0.0.3 && tcp && tcp.dst == 2222`

While solution 2 would have 2 port groups:
- port group X1 would contain only Pod A and would have a single ACL `ip.src == {Z} && tcp && tcp.dst == 1111`
- port group X2 would contain Pod B and Pod C and would have a single ACL `ip.src == {Z} && tcp && tcp.dst == 2222`

Egress policies would require splitting up the ACLs in a similar way, but would probably require a completely separate set of code, because in egress policies, the `namespacePolicy`'s port group represents the source pods, but the `ports` apply to the destinations, not the source.

**Comment by danwinship** [comment]
(FTR, it is my opinion that this feature should not exist. I tried to get it killed once but failed.)

**Comment by trozet** [comment]
Thanks @danwinship I personally like option 1 the best (and also don't really like this feature either). Let's sync up tmrw just to make sure my understanding is correct and then I can add the missing logic.

**Comment by trozet** [comment]
After discussing with @danwinship and @dcbw we will add "basic" named port support. This support means that if someone uses a named port, we will look up in the relevant namespace all of the named ports applicable, but we will only use the first one. This way we can avoid having to refactor and complicate the ACL code we create for this. In the event that there are more than 1 named port value used for the same named port in a namespace, we will issue a warning using the event recorder for the policy.

@girishmg FYI ^

**Comment by danwinship** [comment]
> Caveats include:
> 
>  - Support for namespace label selector only (not pod selector)
>  - Found named ports are not updated (network policy update events, pod
>    update/deletion, etc)
>  - If more than 1 named port value exists for qualifying pods, only the
>    first is used. A warning event is written to the network policy to
>    indicate ovn-k8s only supports a single named port per namespace.

That's not "named port support", that's "unskipping the named port test" support. Don't do that. Unskipping tests is not a goal in itself. If your goal is not to actually implement the feature at all, then leave the test skipped. (And to be clear, I don't think you should actually implement the feature.)

(FTR, the implementation is wrong anyway; if you have a policy that matches two pods, A and B, and says to allow traffic on port "http", and pod A defines that as port 80, and pod B doesn't define it, then the policy says to allow traffic to port 80 _on pod A only_, whereas your code will allow to port 80 on both A and B.)

**Comment by trozet** [comment]
@danwinship OK then I'm just going to update the PR to skip named port tests and forget about adding the functionality now. It's not worth the amount of effort to implement it fully.

**Comment by trozet** [comment]
Failing test is:
should allow ingress access from updated pod

 due to a bug in the test case and fixed: 
https://github.com/kubernetes/kubernetes/commit/55b687054d560125819e805be8eace67c9399e1f

Will skip it for now.

**Pull request merged**